{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8bc7dc6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import magic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dafe90f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# file_path = \"../processed_data/pbmc10k_SEACells_summary.h5ad\"\n",
    "file_path = \"../processed_data/GSE130973_SEACells_summary.h5ad\"\n",
    "\n",
    "adata = sc.read(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1df5395",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating MAGIC...\n",
      "  Running MAGIC on 15457 cells and 30867 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/genos/.local/lib/python3.8/site-packages/magic/magic.py:425: UserWarning: Input matrix contains unexpressed genes. Please remove them prior to running MAGIC.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Calculated PCA in 13.45 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 9.07 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 8.22 seconds.\n",
      "  Calculated graph and diffusion operator in 30.75 seconds.\n",
      "  Running MAGIC with `solver='exact'` on 30867-dimensional data may take a long time. Consider denoising specific genes with `genes=<list-like>` or using `solver='approximate'`.\n",
      "  Calculating imputation...\n",
      "  Calculated imputation in 87.27 seconds.\n",
      "Calculated MAGIC in 119.42 seconds.\n"
     ]
    }
   ],
   "source": [
    "magic_op = magic.MAGIC()\n",
    "approx_magic = magic_op.fit_transform(adata, genes='all_genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0bf1231d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(approx_magic, n_top_genes=2000, subset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "71174c44",
   "metadata": {},
   "outputs": [],
   "source": [
    "#approx_magic umap\n",
    "sc.tl.pca(approx_magic, n_comps=100, use_highly_variable=True)\n",
    "\n",
    "approx_magic.obsm['X_pca'] = approx_magic.obsm['X_pca'][:, 0:30] # after getting optimal number of PCs\n",
    "approx_magic.varm['PCs'] = approx_magic.varm['PCs'][:, 0:30]\n",
    "sc.pp.neighbors(approx_magic, n_neighbors=30, use_rep='X_pca', metric='euclidean', key_added='neighbors_30')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "47caabd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(approx_magic, neighbors_key='neighbors_30', min_dist=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f887f9a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata.layers['MAGIC'] = approx_magic.X\n",
    "adata.obsm['X_magic_pca'] = approx_magic.obsm['X_pca']\n",
    "adata.obsm['X_magic_umap'] = approx_magic.obsm['X_umap']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e1cf80c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers['log_norm'] = adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ed73c98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_is_filtered</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_reference</th>\n",
       "      <th>feature_biotype</th>\n",
       "      <th>feature_length</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>True</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>623</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>True</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>888</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>True</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>2618</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>629</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>1.753778e-03</td>\n",
       "      <td>-4.546900</td>\n",
       "      <td>-1.053845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1319</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000212907</th>\n",
       "      <td>False</td>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>297</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>5.563127e-01</td>\n",
       "      <td>-1.459180</td>\n",
       "      <td>-0.092909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <td>False</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1378</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>2.956016e+00</td>\n",
       "      <td>1.260410</td>\n",
       "      <td>0.000050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <td>False</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1812</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>1.236362e+00</td>\n",
       "      <td>0.019244</td>\n",
       "      <td>-0.362547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>False</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>525</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>6.324706e-02</td>\n",
       "      <td>-3.498859</td>\n",
       "      <td>0.004616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <td>False</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1141</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>2.673762e+00</td>\n",
       "      <td>1.068565</td>\n",
       "      <td>-0.093927</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30867 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 feature_is_filtered     feature_name feature_reference  \\\n",
       "ENSG00000243485                 True      MIR1302-2HG    NCBITaxon:9606   \n",
       "ENSG00000237613                 True          FAM138A    NCBITaxon:9606   \n",
       "ENSG00000186092                 True            OR4F5    NCBITaxon:9606   \n",
       "ENSG00000238009                False  ENSG00000238009    NCBITaxon:9606   \n",
       "ENSG00000239945                 True  ENSG00000239945    NCBITaxon:9606   \n",
       "...                              ...              ...               ...   \n",
       "ENSG00000212907                False          MT-ND4L    NCBITaxon:9606   \n",
       "ENSG00000198886                False           MT-ND4    NCBITaxon:9606   \n",
       "ENSG00000198786                False           MT-ND5    NCBITaxon:9606   \n",
       "ENSG00000198695                False           MT-ND6    NCBITaxon:9606   \n",
       "ENSG00000198727                False           MT-CYB    NCBITaxon:9606   \n",
       "\n",
       "                feature_biotype feature_length    feature_type  \\\n",
       "ENSG00000243485            gene            623          lncRNA   \n",
       "ENSG00000237613            gene            888          lncRNA   \n",
       "ENSG00000186092            gene           2618  protein_coding   \n",
       "ENSG00000238009            gene            629          lncRNA   \n",
       "ENSG00000239945            gene           1319          lncRNA   \n",
       "...                         ...            ...             ...   \n",
       "ENSG00000212907            gene            297  protein_coding   \n",
       "ENSG00000198886            gene           1378  protein_coding   \n",
       "ENSG00000198786            gene           1812  protein_coding   \n",
       "ENSG00000198695            gene            525  protein_coding   \n",
       "ENSG00000198727            gene           1141  protein_coding   \n",
       "\n",
       "                 highly_variable         means  dispersions  dispersions_norm  \n",
       "ENSG00000243485            False  1.000000e-12          NaN               NaN  \n",
       "ENSG00000237613            False  1.000000e-12          NaN               NaN  \n",
       "ENSG00000186092            False  1.000000e-12          NaN               NaN  \n",
       "ENSG00000238009            False  1.753778e-03    -4.546900         -1.053845  \n",
       "ENSG00000239945            False  1.000000e-12          NaN               NaN  \n",
       "...                          ...           ...          ...               ...  \n",
       "ENSG00000212907            False  5.563127e-01    -1.459180         -0.092909  \n",
       "ENSG00000198886            False  2.956016e+00     1.260410          0.000050  \n",
       "ENSG00000198786            False  1.236362e+00     0.019244         -0.362547  \n",
       "ENSG00000198695            False  6.324706e-02    -3.498859          0.004616  \n",
       "ENSG00000198727            False  2.673762e+00     1.068565         -0.093927  \n",
       "\n",
       "[30867 rows x 10 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8523af0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the modified AnnData object\n",
    "# adata.write('../processed_data/pbmc10k_SEACells_MAGIC.h5ad', compression='gzip')\n",
    "adata.write('../processed_data/GSE130973_SEACells_MAGIC.h5ad', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb342d8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã— n_vars = 15457 Ã— 2000\n",
       "    obs: 'age', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'donor_id', 'suspension_type', 'Cluster', 'Celltype', 'tissue_type', 'sample_id', 'library_id', 'library_preparation_batch', 'library_sequencing_run', 'alignment_software', 'manner_of_death', 'sample_source', 'sample_collection_method', 'institute', 'sampled_site_condition', 'sample_preservation_method', 'sequenced_fragment', 'reference_genome', 'cell_enrichment', 'gene_annotation_version', 'cell_type', 'assay', 'disease', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'SEACell'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'feature_type', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'hvg', 'pca', 'neighbors_30', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'neighbors_30_distances', 'neighbors_30_connectivities'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "approx_magic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f5fe161a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_is_filtered</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_reference</th>\n",
       "      <th>feature_biotype</th>\n",
       "      <th>feature_length</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>False</td>\n",
       "      <td>HES4</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>961</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.385177</td>\n",
       "      <td>0.761288</td>\n",
       "      <td>2.028188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>789</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.268602</td>\n",
       "      <td>-0.442180</td>\n",
       "      <td>3.098512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <td>False</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1039</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.120312</td>\n",
       "      <td>-0.838435</td>\n",
       "      <td>2.697751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000179403</th>\n",
       "      <td>False</td>\n",
       "      <td>VWA1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1483</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.048009</td>\n",
       "      <td>-1.430383</td>\n",
       "      <td>2.099072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197530</th>\n",
       "      <td>False</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1455</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.157099</td>\n",
       "      <td>-1.569897</td>\n",
       "      <td>1.957971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160180</th>\n",
       "      <td>False</td>\n",
       "      <td>TFF3</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>452</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.619089</td>\n",
       "      <td>3.555539</td>\n",
       "      <td>3.606666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160191</th>\n",
       "      <td>False</td>\n",
       "      <td>PDE9A</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1724</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.015172</td>\n",
       "      <td>-1.810278</td>\n",
       "      <td>1.714856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000183255</th>\n",
       "      <td>False</td>\n",
       "      <td>PTTG1IP</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>755</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.238921</td>\n",
       "      <td>-1.989137</td>\n",
       "      <td>1.533963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160255</th>\n",
       "      <td>False</td>\n",
       "      <td>ITGB2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>692</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.164043</td>\n",
       "      <td>-0.086117</td>\n",
       "      <td>3.458624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160307</th>\n",
       "      <td>False</td>\n",
       "      <td>S100B</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>971</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>0.076188</td>\n",
       "      <td>0.224100</td>\n",
       "      <td>3.772370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 feature_is_filtered feature_name feature_reference  \\\n",
       "ENSG00000188290                False         HES4    NCBITaxon:9606   \n",
       "ENSG00000186891                False     TNFRSF18    NCBITaxon:9606   \n",
       "ENSG00000186827                False      TNFRSF4    NCBITaxon:9606   \n",
       "ENSG00000179403                False         VWA1    NCBITaxon:9606   \n",
       "ENSG00000197530                False         MIB2    NCBITaxon:9606   \n",
       "...                              ...          ...               ...   \n",
       "ENSG00000160180                False         TFF3    NCBITaxon:9606   \n",
       "ENSG00000160191                False        PDE9A    NCBITaxon:9606   \n",
       "ENSG00000183255                False      PTTG1IP    NCBITaxon:9606   \n",
       "ENSG00000160255                False        ITGB2    NCBITaxon:9606   \n",
       "ENSG00000160307                False        S100B    NCBITaxon:9606   \n",
       "\n",
       "                feature_biotype feature_length    feature_type  \\\n",
       "ENSG00000188290            gene            961  protein_coding   \n",
       "ENSG00000186891            gene            789  protein_coding   \n",
       "ENSG00000186827            gene           1039  protein_coding   \n",
       "ENSG00000179403            gene           1483  protein_coding   \n",
       "ENSG00000197530            gene           1455  protein_coding   \n",
       "...                         ...            ...             ...   \n",
       "ENSG00000160180            gene            452  protein_coding   \n",
       "ENSG00000160191            gene           1724  protein_coding   \n",
       "ENSG00000183255            gene            755  protein_coding   \n",
       "ENSG00000160255            gene            692  protein_coding   \n",
       "ENSG00000160307            gene            971  protein_coding   \n",
       "\n",
       "                 highly_variable     means  dispersions  dispersions_norm  \n",
       "ENSG00000188290             True  0.385177     0.761288          2.028188  \n",
       "ENSG00000186891             True  0.268602    -0.442180          3.098512  \n",
       "ENSG00000186827             True  0.120312    -0.838435          2.697751  \n",
       "ENSG00000179403             True  0.048009    -1.430383          2.099072  \n",
       "ENSG00000197530             True  0.157099    -1.569897          1.957971  \n",
       "...                          ...       ...          ...               ...  \n",
       "ENSG00000160180             True  0.619089     3.555539          3.606666  \n",
       "ENSG00000160191             True  0.015172    -1.810278          1.714856  \n",
       "ENSG00000183255             True  0.238921    -1.989137          1.533963  \n",
       "ENSG00000160255             True  0.164043    -0.086117          3.458624  \n",
       "ENSG00000160307             True  0.076188     0.224100          3.772370  \n",
       "\n",
       "[2000 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "approx_magic.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "09694647",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata.X as pd.DataFrame\n",
    "X = pd.DataFrame(adata.X.toarray(), index=adata.obs.index, columns=adata.var.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1dbd98b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"â–¸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"â–¾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>MAGIC(knn_max=15)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">MAGIC</label><div class=\"sk-toggleable__content\"><pre>MAGIC(knn_max=15)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "MAGIC(knn_max=15)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# iterate for genes 'ENSG00000160191' and 'ENSG00000186827' until t = 5\n",
    "magic_op"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sc_rna_env_python",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
